clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -859,53 +859,60 @@ CodeGenFunction::emitBuiltinObjectSize(const Expr *E, unsigned Type,
   }
 
   if (IsDynamic) {
-    LangOptions::StrictFlexArraysLevelKind StrictFlexArraysLevel =
-        getLangOpts().getStrictFlexArraysLevel();
-    const Expr *Base = E->IgnoreParenImpCasts();
-
-    if (FieldDecl *FD = FindCountedByField(Base, StrictFlexArraysLevel)) {
-      const auto *ME = dyn_cast<MemberExpr>(Base);
-      llvm::Value *ObjectSize = nullptr;
-
-      if (!ME) {
-        const auto *DRE = dyn_cast<DeclRefExpr>(Base);
-        ValueDecl *VD = nullptr;
-
-        ObjectSize = ConstantInt::get(
-            ResType,
-            getContext().getTypeSize(DRE->getType()->getPointeeType()) / 8,
-            true);
-
-        if (auto *RD = DRE->getType()->getPointeeType()->getAsRecordDecl())
-          VD = RD->getLastField();
-
-        Expr *ICE = ImplicitCastExpr::Create(
-            getContext(), DRE->getType(), CK_LValueToRValue,
-            const_cast<Expr *>(cast<Expr>(DRE)), nullptr, VK_PRValue,
-            FPOptionsOverride());
-        ME = MemberExpr::CreateImplicit(getContext(), ICE, true, VD,
-                                        VD->getType(), VK_LValue, OK_Ordinary);
-      }
-
-      // At this point, we know that \p ME is a flexible array member.
-      const auto *ArrayTy = getContext().getAsArrayType(ME->getType());
+    // The code generated here calculates the size of a struct with a flexible
+    // array member that uses the counted_by attribute. There are two instances
+    // we handle:
+    //
+    //       struct s {
+    //         unsigned long flags;
+    //         int count;
+    //         int array[] __attribute__((counted_by(count)));
+    //       }
+    //
+    //   1) bdos of the flexible array itself:
+    //
+    //     __builtin_dynamic_object_size(p->array, 1) ==
+    //         p->count * sizeof(*p->array)
+    //
+    //   2) bdos of the whole struct, including the flexible array:
+    //
+    //     __builtin_dynamic_object_size(p, 1) ==
+    //        sizeof(*p) + p->count * sizeof(*p->array)
```

Comment:
Due to the alignment and padding, the size of `struct s` isn't always the start of array. FYI https://godbolt.org/z/bMb19n9Tz

I think the equation should be: 
`max (offset(struct s, array) + p->count * sizeof(*p->array), sizeof(*p))`

Reply:
I'm not sure I understand the issue involved here. From what I can tell, the `getContext().getTypeSize(StructTy)` returns the number of bits in the whole structure, which I divide by 8 to get bytes.

Reply:
This is about the semantics.

`int array[__counted_by(count)]` means you have `count` elements starting right from the flexible member `array` offset, which is `offsetof(struct s, array)`.


Whereas, `sizeof(*p) + p->count * sizeof(*p->array)` indicates you have `count` elements starting from at the end of the struct, which isn't always the same as the offset of `array` because of the padding and alignment rule.

Reply:
But I never want `offsetof` here. The `sizeof(*p)` will report the correct size of the array without adding on any size for the flexible array member, when it has no value. So `sizeof(*p) for the array in the example is `16`, which is correct. Possibly the only issue may be when the flexible array member isn't *strict*, i.e. it has something like `int array[0] or int array[1]`.

Reply:
That was why I suggested to have `max` here: `max (offset(struct s, array) + p->count * sizeof(*p->array), sizeof(*p))` so that it can return the correct size when the struct size is bigger than the end of the array due to the padding.

Reply:
```
struct flex {
    double dummy;
    char c;
    char fam[__counted_by(c)];
};
```

[0~8)      : double dummy
[8~9)      : char c = 7
[9~ 16) : char fam[__counted_by(c)];

In the above case, I think `__builtin_dynamic_object_size` should return `16`, but `sizeof(*p) + p->count * sizeof(*p->array)` will return `16 + 7 = 23`.

Reply:
I just tested it and both of these produce the correct `sizeof` for the struct:

```
struct flex {
  double dummy;
  char c;
  char fam[0] __attribute__((counted_by(c)));
};

struct flex {
  double dummy;
  char c;
  char fam[1] __attribute__((counted_by(c)));
};
```

even when `-fstrict-flex-arrays=0` is specified. However, if I do this:

```
struct flex {
  double dummy;
  char c;
  char fam[42] __attribute__((counted_by(c)));
};
```
then `-fstrict-flex-arrays=0` reports `sizeof(struct flex)` to include the 42 `fam` elements. This is unfortunate, and I'll have to place a check in there for it.

Thank you for bringing this up!

Reply:
Note: `sizeof(*p) + p->count * sizeof(*p->array)` returning `23` is exactly what I want here. :-) At least with regard to `__builtin_dynamic_object_size()`.

Reply:
> Note: sizeof(*p) + p->count * sizeof(*p->array) returning 23 is exactly what I want here. :-) At least with regard to __builtin_dynamic_object_size().

hmm, I see. So, that's not what `-fbounds-safety` is meant to work: `__counted_by` on fam in -fbounds-safety is to exactly add `count` trailing elements from `&s->fam[0]`. So I think this semantic difference between two modes going to be a problem (e.g., accessing the offset 22 of the object, for example, will be considered OOB in -fbounds-safety).

Is there a reason you want it to be this way? I know people tends to do `malloc(sizeof(struct s) + p->count * sizeof(int));` to allocate an object with fam. But people also use `offsetof` and in such cases, `__builtin_dynamic_object_size` can be bigger than what it should be.

```
struct flex {
  double dummy;
  char c;
  char fam [__counted_by(7)];
};
```

 If `__builtin_dynamic_object_size` does `sizeof(*p) + p->count * sizeof(*p->array)`, it's like effectively considering `&s->fam[13]` as part of the object.

Reply:
I’ve seen `sizeof(*p) + p->count * sizeof(*p->array)` often as the argument to `malloc`, which isn’t terribly problematic because it over-allocates a little bit in the worst case. However, it _would_ be a bug to use it for virtually anything else, such as the size of `memcpy` or `memset`, because it will cause an out-of-bounds access if the object hasn’t been allocated with `malloc`. Both clang and gcc allow structs with flexible array members to be allocated with static storage duration. See this ASAN failure: https://godbolt.org/z/Gxbv379qE

As Yeoul said, the correct end address is calculated with `&fam->array[fam->count]`. Ahead of time, `offset` can be used to compute the required size of a structure with a given number of flexible elements: `offsetof(struct flex, .array[count])`.

Reply:
`__builtin_dynamic_object_size` only adds the full struct size to the calculation when the full struct pointer is specified: `__builtin_dynamic_object_size(p, 1) == 23`. When it's specified on the fam itself (`__builtin_dynamic_object_size(*p->array, 1)`) it returns only the size of the fam (7 in this example). This seems entirely reasonable to me and fits the definition of `__builtin_dynamic_object_size`:

> is similar to __builtin_object_size in that it returns a number of bytes from ptr to the end of the object ptr pointer points to, except that the size returned may not be a constant.

Reply:
> __builtin_dynamic_object_size only adds the full struct size to the calculation when the full struct pointer is specified: __builtin_dynamic_object_size(p, 1) == 23. __builtin_dynamic_object_size only adds the full struct size to the calculation when the full struct pointer is specified: __builtin_dynamic_object_size(p, 1) == 23. When it's specified on the fam itself (__builtin_dynamic_object_size(*p->array, 1)) it returns only the size of the fam (7 in this example). This seems entirely reasonable to me and fits the definition of __builtin_dynamic_object_size:

I know, but the question is why "the full struct size" should include the part of `&s->fam[13]`? It's not even conformed to how the statically initialized struct size is determined in C (like @apple-fcloutier 's example also indicates).

Reply:
I cannot agree that `sizeof(*p) + p->count * sizeof(*p->array)` is the reasonable definition of size when ASAN can diagnose it as out of bounds.

Reply:
Why wouldn't it include the FAM part if the full struct pointer is specified to __bdos? This usage is something that shows up in the Linux kernel (a motivating factor for this feature).

As for @apple-fcloutier 's example, The compiler would complain that `__counted_by(c)` is applied to a non-flexible array member. If we want that capability, we'll have to add it in at a later date.

Reply:
@apple-fcloutier again, why? I *still* don't see the reason why this would be confusing or cause erroneous checks.

Reply:
The structure is defined with a flexible array member:

```c
struct flex {
    double dummy;
    char c;
    char fam[/* __counted_by(c) */];
};
```

Reply:
> Why wouldn't it include the FAM part if the full struct pointer is specified to __bdos? This usage is something that shows up in the Linux kernel (a motivating factor for this feature).

I meant the full struct size should include the FAM part that is specified by `__counted_by(7)`, but not more than that (e.g., `&s->fam[13]`).

Reply:
@rapidsna That's what this feature does. This is why I'm so confused by this argument. :-)

Reply:
@apple-fcloutier Ah, I didn't see the square brackets. So this only works if the `counted_by` attribute is specified on the fam and set to the correct number. In the documentation, it's explicitly stated that it's up to the programmer to ensure that the count is correct if the size somehow changes.

Reply:
> As for @apple-fcloutier 's example, The compiler would complain that `__counted_by(c)` is applied to a non-flexible array member. If we want that capability, we'll have to add it in at a later date.

I think we might have to because some structs are both statically and dynamically initialized. But the point the example was trying to make was to show how struct sizes with trailing arrays are normally calculated.

Reply:
@rapidsna I misread the example. The `__counted_by` attr can exist there and the value of `c` is set correctly, so `__bdos(p, 1)` should return the correct value (18 in this case).

Reply:
> @rapidsna That's what this feature does. This is why I'm so confused by this argument. :-)

Sorry, I'm a bit confused. Could you clarify what you meant by "what this feature does"?

Reply:
> Sorry, I'm a bit confused. Could you clarify what you meant by "what this feature does"?

Responding to:

> I meant the full struct size should include the FAM part that is specified by __counted_by(7), but not more than that (e.g., &s->fam[13]).

So `s->fam[13]` access should fail, of course. And so we use `__bdos(p->array, 1)` to get the size of the FAM. It can then be checked against this index. There is code however that passes in the struct pointer itself to `__bdos(p, 1)`. In those cases, we also add in the struct size. The assumption is that the user really did mean to calculate the full size (struct + fam). Before this feature, that `__bdos` call would return `-1`. Now it can return something that allows for better array access checking.

Reply:
Sorry, I don’t understand what we don’t mutually understand. In the same situation:

```
#include <string.h>

struct flex {
    double dummy;
    char c;
    char fam[__counted_by(c)];
};

struct flex f = { 123.0, 2, { 1, 2 } };

int main() {
    memset(&f, 0, sizeof(f) + f.c * sizeof(*f.fam));
}
```

We have an ASAN trap. Here are all the ways I can think of to reconcile this fact with the model that you propose:

* 2 is the wrong count value for a 2-element array;
* Clang is under-allocating for `f`;
* this is in-bounds and ASAN has a false positive;
* it's actually OK to write out of bounds when relying on `__builtin_dynamic_object_size(p, 1)`;
* the definition chosen for `__builtin_dynamic_object_size` is wrong.

The only one that makes sense to me is that the definition chosen for `__builtin_dynamic_object_size`. Which one is it to you? Do you see another way out?

Reply:
Okay, I think we are talking past to each other a little bit. That comment I was responding to this:

> Why wouldn't it include the FAM part if the full struct pointer is specified to __bdos? 

I'm saying "the full struct size" isn't exactly `struct + fam` because fam doesn't always exactly start from sizeof(struct) when there is a padding in the struct due to the alignment. 

As we all see in the previous examples, the full struct size is `offsetof(struct s, fam) + sizeof(*p->array) * p->count` and align the result to `alignof(struct s)`.

Reply:
@rapidsna Ah! Okay. That makes more sense, thank you. I'll see what I can do here.

Reply:
Note that it's okay for the `count * sizeof(fam)` to be less than the actual size of the fam. It's there to ensure the index doesn't go over what's expected.

Reply:
Okay, I added dc8f0df, which should take care of this issue here. Sorry about my confusion. PTAL.